### ISSUE TITLE
calculate coordinates dialog doesn't open with long formula in clipboard

### ISSUE THREAD

**Issue by ztNFny** [cgeo/cgeo#13994]
### Describe your problem!

Calculate coordinates dialog gets killed - i guess it's trying to parse a long formula and runnign into problems. maybe the "x"?
@eddiemuc 

### How to reproduce?

1. Put a long formula in clipboard:
N 50° 00.A + B + C + D + E + F + G + H + I + J + K + L + M
E 009° 00.A x B x C x ( N + O ) + P + Q + R + S - T - U - V - W

2. In a cache create a new waypoint and select "Calculate Coordinate"
-> c:geo is unresponsive for some time, then stops

02-19 14:19:25.562  2921 24411 E ActivityManager: ANR in cgeo.geocaching (cgeo.geocaching/.EditWaypointActivity)
02-19 14:19:25.562  2921 24411 E ActivityManager: PID: 21443
02-19 14:19:25.562  2921 24411 E ActivityManager: Reason: Input dispatching timed out (8ee85b7 cgeo.geocaching/cgeo.geocaching.EditWaypointActivity (server) is not responding. Waited 5005ms for FocusEvent(hasFocus=true))
02-19 14:19:25.562  2921 24411 E ActivityManager: Parent: cgeo.geocaching/.EditWaypointActivity
02-19 14:19:25.562  2921 24411 E ActivityManager: Frozen: false

### Reproducible

Yes

### c:geo Version

2023.02.18

**Comment by eddiemuc** [comment]
That's really weird. It seems the find-method of the Android-specific version of the java.util.Matcher class is stuck in an endless loop when this specific expression is parsed. This does not happen when the Matcher class from original Java SDK is used (hence the error does not occur in unit tests, but only if I move the test to androidTest).

**Comment by eddiemuc** [comment]
Here it stucks:

```
"Instr: androidx.test.runner.AndroidJUnitRunner@20407" prio=5 tid=0x3d nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at com.android.icu.util.regex.MatcherNative.findImpl(MatcherNative.java:-1)
	  at com.android.icu.util.regex.MatcherNative.find(MatcherNative.java:81)
	  at java.util.regex.Matcher.find(Matcher.java:579)
	  - locked <0x5790> (a java.util.regex.Matcher)
	  at cgeo.geocaching.utils.formulas.FormulaUtils.scanCoordinatePattern(FormulaUtils.java:202)
	  at cgeo.geocaching.utils.formulas.FormulaUtils.scanForCoordinates(FormulaUtils.java:177)
	  at cgeo.geocaching.utils.InstrumentedFormulaUtilsTest.assertScanCoordinates(InstrumentedFormulaUtilsTest.java:208)
	  at cgeo.geocaching.utils.InstrumentedFormulaUtilsTest.scanLongText(InstrumentedFormulaUtilsTest.java:203)
	  at java.lang.reflect.Method.invoke(Method.java:-1)
	  at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	  at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	  at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	  at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	  at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	  at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	  at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	  at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	  at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	  at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	  at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	  at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	  at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	  at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	  at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	  at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	  at org.junit.runners.Suite.runChild(Suite.java:128)
	  at org.junit.runners.Suite.runChild(Suite.java:27)
	  at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	  at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	  at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	  at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	  at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	  at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	  at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	  at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	  at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	  at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:67)
	  at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:58)
	  at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:446)
	  at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2248)
```

**Comment by moving-bits** [comment]
Got a similar report on support today (ticket 104368):

---

I noticed the following (bug):
When I got the following string in the clipboard (without ""):
"N 51° 14.000 + A + C + E + G + I + K + M + O"

and try to "Add waypoint" in a cache and then tab on the Coordinates and then "(x) Calculate coordinates" the menu doesn't open. Instead I get a message that c:geo is not responding after tabbing around in the app.

If the clipboard contains some other content (even parts of the above) or is empty "(x) Calculate coordinates" opens as expected.

**Comment by Lineflyer** [comment]
@ztNFny Did you also experience, that you received permanent ANR notifications afterwards (as I described in #13995)?

**Comment by ztNFny** [comment]
> @ztNFny Did you also experience, that you received permanent ANR notifications afterwards (as I described in #13995)?

No, the Edit Waypoint dialog is killed by ActivityManager, then it's back to normal

**Comment by eddiemuc** [comment]
Ok, I learned something about Android: it provides its OWN implementation of Pattern + Matcher with its OWN little set of bugs and problems.... In current case, Android's implementation chose to freeze on certain (valid) patterns when trying to match them.

This, in conjunction with a scan of clipboard on opening of calculated coordinate dialog (to decide whether to enable/disable the "copy coordinates from clipboard" action), lead to freeze of dialog.

Fixed with PR #13996. This PR also introduces certain scan tests as instrumented tests (those use the Android Pattern/Matcher implementation, while the normal unit tests use the normal Java JRE ones)
